server/src/code-runner/docker-runner.ts

```diff
@@ -91,16 +123,9 @@ export class DockerRunner implements CodeRunner {
               })
             ),
             concat(this.config.downloader.fetch(containerId, configuration.inputs)),
-            concat(
-              this.config.spawn(this.config.dockerExecutable, [
-                'exec',
-                '-t',
-                containerId,
-                '/bin/bash',
-                '-c',
-                `mkdir /run/outputs && cd /run && python main.py ${args}`
-              ])
-            ),
+            // Only install `requirements.txt` when the container is writeable
+            concat(this.runCommand(containerId, 'cd /run && test -f requirements.txt && pip install -r requirements.txt', environment.writeable)),
```

Comment:
Instead of a hardcoded `pip install` this should rather execute an arbitrary, user-provided (think: `ml.yaml`) pre-execution command. That way, we are flexible regarding other ecosystems such as running a `yarn install`. We can still add an automatic environment detection that then just facilitates the existing lower level `pre-execution` hook.

Reply:
Any suggestions for the property in `ml.yaml`? Something like [Travis](https://docs.travis-ci.com/user/customizing-the-build) does with `before_install`? But then `beforeInstall` which could have an array of commands and are chained together with `&&`?

Reply:
Yes, that sounds about right. I'd vote for calling that property `pre_execution` or `before_execution`. Or maybe even `before_execution_run` which could go well with an `execution_run` or `run_execution` property which could specify things like `node foo.js` etc.

Reply:
I think we should stick with `camelCase` because all other properties are also named like that.

I'm fine with something like `beforeExecution` or `beforeRun`.

Reply:
Let me throw in `preExecutionCmd`

Reply:
Then I would go with `preExecutionCommand` :D